src/cache.rs

```diff
@@ -0,0 +1,152 @@
+use std::any::Any;
+use std::collections::HashMap;
+use std::error::Error;
+use std::sync::RwLock;
+use std::thread;
+use std::time::Instant;
+
+type Generator = fn() -> Result<Box<Any>, Box<Error>>;
+
+const CACHE_TTL_SECS: u64 = 120;
+
+lazy_static! {
+    static ref CACHE: RwLock<HashMap<Generator, (Box<Any + Send + Sync>, Instant)>> =
+        RwLock::new(HashMap::new());
+}
+
+pub fn get<T>(generator: Generator) -> Result<T, Box<Error>>
+where
+    T: Send + Sync + Clone + 'static,
+{
+    if let Some(cached) = get_cached(generator) {
+        Ok(cached)
+    } else {
+        update_cache(generator)
+    }
+}
+
+fn get_cached<T>(generator: Generator) -> Option<T>
+where
+    T: Send + Sync + Clone + 'static,
+{
+    let cache = CACHE.read().unwrap();
+    cache
+        .get(&generator)
+        .map(|&(ref data, timestamp)| {
+            if timestamp.elapsed().as_secs() > CACHE_TTL_SECS {
+                // Update the cache in the background
+                thread::spawn(move || {
+                    let _ = update_cache::<T>(generator);
+                });
+            }
+            data
+        })
+        .and_then(|data| data.downcast_ref::<T>().map(|t| t.clone()))
```

Comment:
I think unwrap/clone would make more sense here -- if this isn't a T that's a bug in the code, not a missing value, correct?

Reply:
Yes, I'll use unwrap.